### PR TITLE
T232510 - Add namespace filter to the interaction endpoint

### PR DESCRIPTION
**Pull request by dayllanmaza**

### CODE REVIEW

File: server/src/Service/InteractionService.php [hunk 1]
```diff
@@ -113,6 +116,18 @@ private function validateLimit( $limit ) {
 		}
 	}
 
+	/**
+	 * @param array $namespaces
+	 * @throws \InvalidArgumentException
+	 */
+	private function validateNamespaces( $namespaces ) {
+		foreach ( $namespaces as $ns ) {
+			if ( !is_numeric( $ns ) ) {
```
Comment (dayllanmaza) [inline server/src/Service/InteractionService.php]: ugg... What's the least ugly way to check that `$ns` is an integer or the string equivalent of one? `(int) $ns == $ns` ?

Comment (davidbarratt) [reply server/src/Service/InteractionService.php]: or [is_int()](https://www.php.net/manual/en/function.is-int.php)

Comment (davidbarratt) [reply server/src/Service/InteractionService.php]: but I think [is_numeric()](https://www.php.net/manual/en/function.is-numeric.php) is what you want.

Comment (dayllanmaza) [reply server/src/Service/InteractionService.php]: `is_numeric()` allows for decimals. `is_int()` returns false for strings representing a number, `cbyte_digit()` fails for negative numbers. `(int) $ns == $ns` is the simplest check I can think :man_shrugging:

Comment (davidbarratt) [reply server/src/Service/InteractionService.php]: would it be better just type typecast it and not check?

Comment (dayllanmaza) [reply server/src/Service/InteractionService.php]: Well I thought of it but then if someone sends a string like "Main" typecasting will make it a 0 and I would need to special case that in there. And I would need to validate anyway 'cause there is no way to enforce int[]

Comment (davidbarratt) [reply server/src/Service/InteractionService.php]: interesting... yeah I guess that makes sense given what you're trying to do... just odd I guess.

File: server/src/Service/InteractionService.php [hunk 2]
```diff
@@ -113,6 +116,18 @@ private function validateLimit( $limit ) {
...
 	}
 
+	/**
+	 * @param array $namespaces
+	 * @throws \InvalidArgumentException
+	 */
+	private function validateNamespaces( $namespaces ) {
+		foreach ( $namespaces as $ns ) {
+			if ( !is_numeric( $ns ) || (int)$ns != $ns ) {
+				throw new \InvalidArgumentException( 'only namespaces ids are allowed' );
```
Comment (Tchanders) [inline server/src/Service/InteractionService.php]: How about 'namespaces must be specified by ID'?

Perhaps ideally we'd get a list of valid namespaces from the wiki to check whether that namespace actually exists, but that would be adding an extra step. (Would be easier if we were doing this from within MediaWiki.)

File: README.md [hunk 3]
```diff
@@ -75,19 +75,20 @@ webservice --backend=kubernetes restart
 ## API Endpoints
 ### Interaction
 ```
-/api/{wiki}/interaction?user={username1|username2}&start_date={timestamp}&end_date={timestamp}&limit={50}&continue={continue_string}
+/api/{wiki}/interaction?user={username1|username2}&namespace={0|1|2}&start_date={timestamp}&end_date={timestamp}&limit={50}&continue={continue_string}
 ```
 ### Parameters
 * `wiki:` Wiki project we want to check the interactions on. Ex: enwiki, eswiki, testwiki
 * `user:` List of usernames involved in the interactions joined by a `|`. No less than 2
+* `namespace:` List of namespaces ids we want the interactions on joined by a `|`
```
Comment (Tchanders) [inline README.md]: Grammar correction: "namespace ids"

**Review by Tchanders — APPROVED** [review]
Looking good, seems to be working. Have suggested a couple of English corrections, but approve otherwise.

File: server/src/Service/InteractionService.php (hunk 2 above)
Comment (dayllanmaza) [reply server/src/Service/InteractionService.php]: Changed the message per your suggestion. Thanks

**Review by Tchanders — APPROVED** [review]

**Pull request merged**